map-data/src/main/java/org/triplea/map/reader/XmlParser.java

```diff
@@ -0,0 +1,154 @@
+  public static class AttributeParser {
+    private final String currentTag;
+    private final Map<String, Consumer<String>> attributeHandlers;
+
+    public AttributeParser addAttributeHandler(
+        final String attributeName, final Consumer<String> attributeHandler) {
+      attributeHandlers.put(attributeName, attributeHandler);
+      return this;
+    }
+
+    public TagParser addChildTagHandler(
+        final String childTagName, final ThrowingRunnable tagHandler) {
+      return new TagParser(currentTag, attributeHandlers, new HashMap<>())
+          .addChildTagHandler(childTagName, tagHandler);
+    }
+
+    public BodyParser addBodyHandler(final Consumer<String> bodyHandler) {
+      return new BodyParser(currentTag, attributeHandlers, bodyHandler);
+    }
+
+    public void parse(final XMLStreamReader xmlStreamReader) throws XMLStreamException {
+      new GenericParser(currentTag, attributeHandlers, Map.of(), body -> {}).parse(xmlStreamReader);
+    }
+  }
+
+  @AllArgsConstructor(access = AccessLevel.PRIVATE)
+  public static class TagParser {
+    private final String currentTag;
+    private final Map<String, Consumer<String>> attributeHandlers;
+    private final Map<String, ThrowingRunnable> childTagHandlers;
+
+    public TagParser addChildTagHandler(
+        final String childTagNaem, final ThrowingRunnable tagHandler) {
+      childTagHandlers.put(childTagNaem, tagHandler);
+      return this;
+    }
+
+    public void parse(final XMLStreamReader streamReader) throws XMLStreamException {
+      new GenericParser(currentTag, attributeHandlers, childTagHandlers, body -> {})
+          .parse(streamReader);
+    }
+  }
+
+  @AllArgsConstructor(access = AccessLevel.PRIVATE)
+  public static class BodyParser {
+    private final String currentTag;
+    private final Map<String, Consumer<String>> attributeHandlers;
+    private Consumer<String> bodyHandler;
+
+    public void parse(final XMLStreamReader streamReader) throws XMLStreamException {
+      new GenericParser(currentTag, attributeHandlers, Map.of(), bodyHandler).parse(streamReader);
+    }
+  }
+
+  @Log
+  @AllArgsConstructor(access = AccessLevel.PRIVATE)
+  public static class GenericParser {
+    private final String currentTag;
+    private final Map<String, Consumer<String>> attributeHandlers;
+    private final Map<String, ThrowingRunnable> childTagHandlers;
+    private Consumer<String> bodyHandler;
+
+    public void parse(final XMLStreamReader streamReader) throws XMLStreamException {
+      boolean endTagReached = false;
+
+      log.info("Processing current tag: " + currentTag);
+
+      if (streamReader.getEventType() == XMLStreamReader.START_ELEMENT) {
+        for (int i = 0, n = streamReader.getAttributeCount(); i < n; i++) {
+          final int index = i;
+
+          final String attributeName = streamReader.getAttributeLocalName(index);
+          final Consumer<String> attributeHandler = attributeHandlers.get(attributeName);
```

Comment:
Design consideration:
Currently the attributes are all processed separately, however I think it would be better to aggregate all of the attributes at this point into a hashmap and pass this map to the handler.
This way it becomes easier to check if all of the semantically important attributes are present (even though I have to admit I don't know how much work the dtd does here for us)

Reply:
Side-note, it's really interesting that it looks like body elements are never used, it's all attributes..

So, sketching out using a map to return the attributes, it would be something like this:

```
        XmlParser.tag(TAG_NAME)
            .addAttributeHandler(attributes -> {
              unitType = attributeMap.get("unitType");
              territory = attributeMap.get("territory");
            })
            .parse(streamReader);
```

vs: 

```
        XmlParser.tag(TAG_NAME)
            .addAttributeHandler("unitType", value -> unitType = value)
            .addAttributeHandler("territory", value -> territory = value)
            .parse(streamReader);
```


The prevailing responsibility of the `map.data.elements` classes is to capture exactly what is in the XML and to just be java objects representing the XML data. The 'dtd' today introduces defaults, it's a question whether these POJOs should add those default values in or not.

Eventually these objects will be passed up a layer to `game-core` and assembled into a GameData. This will probably take place in `GameParser` where instead of parsing the raw XML it'll invoke the `map-data` parser and get back a `Game` object which it can then traverse and assemble into a game data.

A next step after that will be to store the `Game` tag data as a `MapData` object in `GameData`, make that immutable and not part of the save game. This offloads a lot of data from `GameData` and we would persist much less. Then from there the idea is a `GameData` would lose more information and instead be a `MapData` and a list of change objects. If you then replay all of the change objects onto the map data you would then get the latest game state. In this way we only need to persist which map generated the original `MapData` and then persist the series of change objects (so no more java serialization).

Which comes back to, should this layer be responsible for adding in default values (in essence do the work of the DTD which adds defaults)?  WDYT @RoiEXLab ? I'm honestly not sure at this point, I can see it argued both ways where maybe we're not quite exactly modeling the data we get from the map and are abstracting the detail of default values, on the other hand there is merit to exactly modeling the XML. I think when we get to a point where `MapData` is not persisted with save games it will be more important. But even then, the translation of a `MapData` to a `GameState`, AKA `GameData` will still be done; all of question of whether it should be done in game-core or in map-data. The other consideration I suppose is when we write map data. Would we want the map data to add in defaults that game-core does not request to be written?

Also of note though, the SAX parser does not look at the DTD at all. The API for parsing here is super-low level. I saw some benchmarks showing it's actually slower than JAXB, but uses the least amount of memory. Since there is no DOM model created, we just pass through the XML with a cursor, that makes sense it would be low memory. On the plus side that should hopefully really help with our current problems where we do see memory issues loading maps. With a cleaner model though perhaps we'll be able to get faster parser, but not guaranteed and really not the goal. 

Something else to consider is a migration path, where we'll probably be allowing a large "OR" of conditions where we would read data. Let's say for example:
```
<territory name="Poland" />
```

And:
```
<territory>Poland</territory>
```

In theory a person could also write:
```
<territory name="Poland">Poland</territory>
```

The latter is going to be non-deterministic which would win, but the point is the parser is looking at multiple locations for the same data. This is a bit more profound perhaps if we start nesting tags in different ways. In that latter case it would be the job of the GameData assembler to choose the right data if presented with multiple options. For example, something like:
```
<relationships>
  <relationship>
         <type name="Peace" />
         <playerRelationship p1="Polish" p2="Austrians" />
  </relationship>
</relationships>
```

Today the above is modeled as two top level tags:
```
<relationships>
   <relationship type="Peace" player1="Polish" player2="Austrians" />
</relationships>

<relationshipTypes>
   <relationshipType name="Peace" />
</relationshipTypes>
```

Hence, the map-data parser is going to have to look at both places and would return both if both are present (again leaving it to the game-data assembler to decide which one to prefer if any). This does make me think it's more cohesive if the map data reader really does no more than model the existing XML and allow the assembler to then make any further decisions about defaulting missing values.

I would also emphasize that the map-data-reader does not have any semantic knowledge of the data it reads. For example the property keys we read are important, to the map-data-reader it's just a set of property tags and it does not care what's in them. Similar for any special encoded values where we have for example colon delimited player names. I think to this end it's probably more reason for defaulting of missing values to happen in the assembler. Of course it'll then be up to the assembler to do any semantic validation (for example if an attribute value is a wrong value).